flake/dev-shell.nix

```
@@ -25,6 +25,13 @@
             "$@"
         '';
       };
+
+      # The shell should not directly depend on `packages.serve-docs`, because
+      # that'd build the docs before entering the shell. Instead, we want to
+      # build the docs only when running 'serve-docs'.
+      build-and-run-docs = pkgs.writeShellScriptBin "serve-docs" ''
+        nix run .#docs
+      '';
```

Running this script outside the Stylix flake would not work because the `.#docs` resolution fails. To support this use case, the path should not be relative:

```suggestion
      build-and-run-docs = pkgs.writeShellScriptBin "serve-docs" ''
        nix run ${inputs.self}#docs
      '';
```

Note that this change does not work because `inputs` is currently unavailable in this scope and becasue `inputs'.self` results in the `error: attribute 'self' missing` error. This needs to be resolved to get the Nix store path to the Stylix flake.

Note that this causes the `build-and-run-docs` derivation to be rebuilt on new commits, which is not currently the case.

---

> I thought about this when writing the PR, but using `self` doesn't achieve what I wanted:
>
> The issue is `self` will be the in-store copy of the flake, as it was when the devshell was build; _not_ the current git worktree checked out locally.
>
> This means if you enter a devshell, make changes, then go to test the docs; you will not see your changes reflected in the docs until you reload your devshell and re-run the docs server.

Good catch. I did not consider `inputs.self` and `./.` drifting apart.

> Added to the clarifying comment, which now reads:
>
> > The shell should not directly depend on `packages.serve-docs`, because
> > that'd build the docs before entering the shell. Instead, we want to
> > build the docs only when running 'serve-docs'.
> >
> > For a similar reason, we can't use `self` as a reference to the flake:
> > `self` represents the flake as it was when the devshell was evaluated,
> > not the local flake worktree that has possibly been modified since
> > entering the devshell.

Looks great.

---

> I did not consider `inputs.self` and `./.` drifting apart.

Technically `self` and `./.` are _both_ the in-store path; there's actually no way to get the original out-of-store path when evaluating as a flake. This is unlike `nix repl` and non-flake evals, where relative paths are out-of-store (unless copied to store using `"${}"`).

But assuming what you actually meant is the in-store path and the out-of-store worktree, then yes, frustratingly we can't really handle this.